<launch>

  <param name="use_sim_time" value="true"/>

  <!-- 启动 Gazebo 仿真环境（世界 + 物理引擎 + GUI）-->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="paused"     value="false"/>   <arg name="use_sim_time" value="true"/>
    <arg name="gui"        value="true"/>    <arg name="headless"   value="false"/>
    <arg name="debug"      value="false"/>   
  </include>


  <!-- 启动 RViz（使用预定义配置文件）-->
  <node name="rviz" pkg="rviz" type="rviz" output="screen" args="-d $(find 4WIS_Robot)/config/robot.rviz"/>


  <!-- 机器人实例（robot_0）-->
  <group ns="robot_0">

    <!-- 机器人初始位姿与名称参数 -->
    <arg name="robot_name" value="robot_0"/>
    <arg name="x"   value="-2.5"/>   <arg name="y"   value="1"/>
    <arg name="z"   value="0.3"/>   <arg name="yaw" value="1.57"/>

    <!-- 生成 robot_description (URDF) -->
    <param name="robot_description" command="$(find xacro)/xacro '$(find 4WIS_Robot)/urdf/car_4ws_4wd.urdf.xacro' robot_name:='$(arg robot_name)'" />


    <!-- 在 Gazebo 中生成机器人模型  -->
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model" output="screen"
          args="-urdf -x $(arg x) -y $(arg y) -z $(arg z) -Y $(arg yaw) -model $(arg robot_name) -param /$(arg robot_name)/robot_description"/>

    <!-- ros_control 控制系统 -->
    <!-- 加载控制器参数（PID、关节名、4WS 运动学等） -->
    <rosparam file="$(find 4WIS_Robot)/config/multi_mode_4ws_controller.yaml" command="load"/>

    <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher">
    <param name="rate" value="50"/>
  </node>
    <!-- controller_manager 中的控制器 -->
    <node name="controller_spawner" pkg="controller_manager" type="spawner"
        respawn="true" output="screen"
        args="multi_mode_4ws_controller"/>

    <!-- robot_state_publisher tf_prefix 用于多机器人隔离 -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen">
      <param name="tf_prefix" value="$(arg robot_name)"/>
    </node>


    <node pkg="4WIS_Robot" name="pointcloud_filter" type="pointcloud_filter.py" output="screen">

      <param name="target_frame" value="$(arg robot_name)/base_link"/>

      <!-- 车体尺寸（base_link 下） -->
      <param name="car_x_min" value="-0.5"/>
      <param name="car_x_max" value="0.5"/>
      <param name="car_y_min" value="-0.31"/>
      <param name="car_y_max" value="0.31"/>
      <!-- 高度范围（base_link 下） -->
      <param name="car_z_min" value="-0.2"/>
      <param name="car_z_max" value="1"/>
    </node>


    <arg name="scanner" default="velodyne"/>

    <!-- 创建 nodelet manager -->
    <node pkg="nodelet" type="nodelet" 
          name="$(arg scanner)_nodelet_manager" 
          args="manager" />

    <!-- 加载 pointcloud_to_laserscan nodelet -->
    <node pkg="nodelet" type="nodelet" 
          name="pointcloud_to_laserscan" 
          args="load pointcloud_to_laserscan/pointcloud_to_laserscan_nodelet $(arg scanner)_nodelet_manager">
      
      <remap from="cloud_in" to="points_filtered"/>
      <remap from="scan" to="scan"/>

      <param name="target_frame" value="$(arg robot_name)/lidar_link"/>
      <param name="transform_tolerance" value="0.05"/>   <!-- 增加容忍度 -->
      <param name="min_height" value="-0.7"/>       
      <param name="max_height" value="1.0"/>
      <param name="angle_min" value="-3.14159"/>      
      <param name="angle_max" value="3.14159"/>       
      <param name="angle_increment" value="0.0087"/>     <!-- 1度: 2*pi/360 -->
      <param name="scan_time" value="0.1"/>              <!-- Velodyne 10Hz -->
      <param name="range_min" value="0.1"/>              <!-- 避免0值 -->
      <param name="range_max" value="8.0"/>
      <param name="use_inf" value="true"/>               <!-- 推荐用inf -->
      <param name="concurrency_level" value="1"/>
    </node>

    <!-- world坐标变换 -->
    <node pkg="4WIS_Robot" type="world_tf.py" name="world_tf" output="screen">
      <param name="robot_name" value="robot_0"/>
      <param name="base_frame" value="$(arg robot_name)/base_link"/>
    </node>

    <!-- ========== 碰撞检测节点 ========== -->
    <node name="collision_visualizer" pkg="4WIS_Robot" type="collision_pub.py" output="screen">
      <param name="robot_name" value="$(arg robot_name)" />
      <param name="marker_duration" value="0.05" />
      <param name="text_height" value="0.5" />
      <param name="frame_id" value="$(arg robot_name)/base_link" />
    </node>

    <!-- ========== 手柄驱动节点 ========== -->
    <node pkg="joy" type="joy_node" name="joy_node" output="screen">
        <!-- 手柄设备 -->
        <param name="dev" value="/dev/input/js0"/>
        <!-- 发布频率 -->
        <param name="autorepeat_rate" value="100.0"/>
        <!-- 死区（可选） -->
        <param name="deadzone" value="0.05"/>
    </node>

    <!-- ========== joy -> cmd_vel 节点，cmdvel的前缀要根据控制插件调整 ========== -->
    <node pkg="4WIS_Robot" type="joy_controller.py" name="joy_to_cmdvel" output="screen">
        <remap from="joy" to="joy"/>
        <remap from="cmd_vel" to="multi_mode_4ws_controller/cmd_vel"/>
        <remap from="mode_select" to="multi_mode_4ws_controller/mode_select"/>
        <param name="max_linear"  value="2.0"/>
        <param name="max_angular" value="0.523"/>
        <param name="deadzone"    value="0.05"/>
    </node>

  </group>
  
</launch>

